File: 841.keys-and-rooms/solution.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        N = len(rooms)
        visited = [False] * N
        visited[0] = True
        q = [newKey for newKey in rooms[0]]

        remainRooms = N - 1

        while (q):
            curr = q.pop(0)
            if (visited[curr]):
                continue
            visited[curr] = True
            remainRooms -= 1

            if (remainRooms == 0):
                break

            for newKey in rooms[curr]:
                if (visited[newKey]):
                    continue
                q.append(newKey)
        return remainRooms == 0

File: 841.keys-and-rooms/solution.kt
import java.util.Queue

class Solution {
    fun canVisitAllRooms(rooms: List<List<Int>>): Boolean {
        val N = rooms.size
        var seen = BooleanArray(N)
        var remainCnt = N

        val q = LinkedList<Int>()
        q.add(0)
        while(!q.isEmpty() && remainCnt > 0) {
            val curr = q.poll()
            if (seen[curr]) continue
            seen[curr] = true
            remainCnt -= 1

            for (newKey in rooms[curr]) {
                if (seen[newKey]) continue
                q.add(newKey)
            }
        }
        return remainCnt == 0
    }
}
